### DISCUSSION TITLE
Use of `to_function` and `set_initial`

### DISCUSSION THREAD

**Discussion by STaliani** [casadi/casadi#3346] [Q&A] [answered]
Hello everyone,
I'm currently working with MPC with casadi and I learned to use `to_function` method to create a casadi function callable multiple times to solve an optimization problem. 
Anyway I'm struggling with two characteristics of this method:
- It doesn't show failures 
- It prevents the use of `set_initial` function

I created a [simple example](https://gist.github.com/STaliani/f956fee4fc4a263431df0caa56364318) to show what I'm talking about.
You can find there a simple system dynamics and three possible approaches to solve the same optimization problem:
- `optimization_problem_no_set_initial(N:int)->cs.Function` uses `to_function` and works fine with the parameters that I give to it.
- `optimization_problem_no_fuction(N:int, x0:np.ndarray, x_ref:np.ndarray, u_ref:np.ndarray)` doesn't use `to_function` instead it directly solve the optimization problem using `set_initial`.  
The drawback of this approach is that one has to call this function to solve the problem and each time is like redefining the same problem. 
Clearly it can't work when things get more complicated because the definition of the optimization problem can take a long time.
- `optimization_problem_set_initial(N:int)->cs.Function` puts in light the impossibility to use `set_initial` when one is using `opti.to_function`.
In fact it when line 104 is not commented out it arises the following error
```
Traceback (most recent call last):
  File "/home/staliani/Documents/prova_casadi/casadi_function.py", line 102, in <module>
    controller_set_initial = optimization_problem_set_initial(N)
  File "/home/staliani/Documents/prova_casadi/casadi_function.py", line 58, in optimization_problem_set_initial
    opti.set_initial(u[:,k], u_ref[:,k])
  File "/home/staliani/robotology-superbuild/build/install/lib/python3/dist-packages/casadi/casadi.py", line 23962, in set_initial
    return _casadi.Opti_set_initial(self, *args)
TypeError: Wrong number or type of arguments for overloaded function 'Opti_set_initial'.
  Possible C/C++ prototypes are:
    casadi::Opti::set_initial(casadi::MX const &,casadi::DM const &)
    casadi::Opti::set_initial(std::vector< casadi::MX,std::allocator< casadi::MX > > const &)
```
The problem is generated by the fact that a reference is passed to a casadi function through a `opti.parameter` declared in the optimization problem. A parameter is a `casadi::MX` type while `set_initial` needs a numeric type.

I am insisting a lot on the use of `set_initial` because it really speeds up things.
Here a quantitative explanation:
- with `set_initial`
```
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
       nlp_f  |  13.00us (  2.60us)   9.58us (  1.92us)         5
       nlp_g  |   7.00us (  1.40us)   7.69us (  1.54us)         5
  nlp_grad_f  |   9.00us (  1.50us)  10.37us (  1.73us)         6
  nlp_hess_l  |   6.00us (  1.50us)   7.13us (  1.78us)         4
   nlp_jac_g  |  11.00us (  1.83us)  11.33us (  1.89us)         6
       total  |   1.05ms (  1.05ms)   1.04ms (  1.04ms)         1
```
- without `set_initial`
```
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
       nlp_f  |   1.59ms (  2.87us)   1.46ms (  2.64us)       553
       nlp_g  |   1.37ms (  2.46us)   1.31ms (  2.36us)       555
    nlp_grad  | 559.00us (  5.70us) 543.66us (  5.55us)        98
  nlp_grad_f  |   1.76ms (  2.86us)   1.71ms (  2.79us)       613
  nlp_hess_l  |   1.12ms (  2.70us)   1.08ms (  2.60us)       415
   nlp_jac_g  |   1.73ms (  2.82us)   1.72ms (  2.81us)       613
       total  |   1.10ms (  1.10ms)   1.10ms (  1.10ms)         1
```
For this simple example the time difference is negligible but looking at the number of evaluations the advantage is clear.

Do you have any work around for the two mentioned issues?
Can I induce the to_function method to, at least, send warnings?
Is there any possibility to include use `set_initial` with parameters while using method`to_function`?

Thanks in advance for your comments

**Comment by jgillis** [comment, marked as answer]
Honestly, I don't recall adding support for parameters as value in set_initial. In any case you should not use such construct in a loop. Instead, you should declare and use parameters once, and in a loop only set parameter values and/or initial values. That avoids reconstructing the problem.

 Anyhow, you can use decision variables  in the inputs list of to_function: 
`func = opti.to_function('controller', [x0, x_ref, u], [u[:,0]], ['x0','x_ref', 'u_init'], ['u'] )`
failures are controlled through 'error_on_fail' option of solvers. Just add that option when declaring opti.solver

**Reply by STaliani** [reply]
Thank you very much for the support!

**Comment by STaliani** [comment]
My doubts are solved, I hope this can help other people too.

**Comment by itbellix** [comment]
Hi everyone, and sorry for reviving this old thread.

I also expected that `opti.set_initial()` could be used with parameters as arguments, in combination with a subsequent `opti.to_function()` to allow for warm-starting of the resulting NLP (once it is embedded in a CasADi function).
I see that there are other threads related to this (https://github.com/casadi/casadi/discussions/3577), and that the issue related to the underlying functionality is not solved yet (https://github.com/casadi/casadi/issues/2206) 

I would just like to confirm that this is still not possible, is it correct @jgillis?

Is it true that, to date, the way to go about this should be either:
1. resorting to the older `nlpsol`
OR
2. using the `opti` object "as-is", without transforming it into a function, and calling `opti.set_initial()` at runtime (when the initial guesses are known numerically) followed by `opti.solve`
?

I would appreciate any clarification. Thank you very much!
